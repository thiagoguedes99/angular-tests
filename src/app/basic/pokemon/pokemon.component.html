@if (charizard) {
  <h3>{{ charizard?.name | titlecase }}</h3>
  <!-- <h3>{{ charizard?.name }}</h3> -->
  <img [src]="charizard?.sprites?.front_default" [alt]="charizard?.name">
}
@else {
  <h2>Loading...</h2>
}
