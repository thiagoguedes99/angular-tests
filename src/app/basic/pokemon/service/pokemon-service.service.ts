import { HttpClient } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';

import { Observable } from 'rxjs';

import { Pokemon } from '../interfaces/pokemon';

@Injectable()
export class PokemonServiceService {
  http: HttpClient = inject(HttpClient);

  constructor() { }


  getPokemon( id: number ):Observable<Pokemon> {
    return this.http.get<Pokemon>(`https://pokeapi.co/api/v2/pokemon/${ id }`);
  }
}
