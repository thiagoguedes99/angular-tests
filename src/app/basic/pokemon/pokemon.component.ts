import { Component, OnInit, inject } from '@angular/core';
import { TitleCasePipe } from '@angular/common';

import { PokemonServiceService } from './service/pokemon-service.service';
import { Pokemon } from './interfaces/pokemon';

@Component({
  selector: 'app-pokemon',
  standalone: true,
  // imports: [CommonModule],
  imports: [TitleCasePipe],
  providers: [PokemonServiceService],
  templateUrl: './pokemon.component.html',
  styleUrl: './pokemon.component.scss'
})
export class PokemonComponent implements OnInit {
  pokemonService: PokemonServiceService  = inject(PokemonServiceService);

  public charizard?: Pokemon;

  ngOnInit(): void {
    this.pokemonService.getPokemon(6)
      .subscribe( pokemon => {
        this.charizard = pokemon;
      });
  }

}
