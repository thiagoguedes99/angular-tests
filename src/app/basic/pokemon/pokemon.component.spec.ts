import { ComponentFixture, TestBed } from '@angular/core/testing';

import { Observable, of } from 'rxjs';

import { PokemonComponent } from './pokemon.component';
import { PokemonServiceService } from './service/pokemon-service.service';
import { HttpClientTestingModule } from '@angular/common/http/testing';

let mockReturnGetPokemon: any = null

const MockPokemonServiceService = {
  getPokemon: () => of(mockReturnGetPokemon)
}


describe('PokemonComponent', () => {
  let component: PokemonComponent;
  let fixture: ComponentFixture<PokemonComponent>;
  let compiled: HTMLBRElement;


  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [HttpClientTestingModule, PokemonComponent],
      providers: [PokemonServiceService],

      // providers: [{
      //   provide: PokemonServiceService,
      //   useValue: MockPokemonServiceService
      //   // useClass: MockPokemonServiceService
      // }]
    })
    .overrideProvider(PokemonServiceService, { useValue: MockPokemonServiceService})
    .compileComponents();

    fixture = TestBed.createComponent(PokemonComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
    compiled = fixture.nativeElement;
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should to match snapshot', () => {
    expect(compiled).toMatchSnapshot();
  });

  it('should to requets in ngOnInit function', () => {
    const mockPokemon = {
      name: 'mock_pokemon_component2',
      sprites: {
        front_default: 'mock_photo_component2'
      }
    }

    mockReturnGetPokemon = mockPokemon;

    component.ngOnInit();

    expect(component.charizard).toBe(mockPokemon);
  });

  it('should to call getPokemon once in call ngOnInit', () => {
    jest.spyOn(MockPokemonServiceService, 'getPokemon')

    component.ngOnInit();

    expect(MockPokemonServiceService.getPokemon).toHaveBeenCalledTimes(1);
  });

  it('should to update in html after fineshed request', () => {
    const mockPokemon = {
      name: 'mock_pokemon_component_html',
      sprites: {
        front_default: 'mock_photo_component_html'
      }
    }

    mockReturnGetPokemon = mockPokemon;

    component.ngOnInit();
    fixture.detectChanges();

    const h3  = compiled.querySelector('h3');
    const img = compiled.querySelector('img');

    expect( h3?.textContent ).toContain( mockPokemon.name.replace('m', 'M') );
    expect( img?.src ).toBe( `http://localhost/${mockPokemon.sprites.front_default}` );
    expect( img?.alt ).toBe( mockPokemon.name );

  });

  it('should to view loading h2 when not has charizard', () => {
    component.charizard = undefined;

    fixture.detectChanges();

    const h2  = compiled.querySelector('h2');
    console.log(compiled.innerHTML)

    expect( h2?.textContent ).toContain('Loading...');
  });
});
