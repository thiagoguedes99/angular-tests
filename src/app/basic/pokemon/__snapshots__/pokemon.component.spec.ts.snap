// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonComponent should to match snapshot 1`] = `
<div
  id="root1"
>
  <h2>
    Loading...
  </h2>
  
  
</div>
`;
